cirq-core/cirq/devices/thermal_noise_model.py

```diff
@@ -0,0 +1,251 @@
+                        out[qb] = 0.0
+                return out
+            else:
+                return {qb: rate_or_dict for qb in qubits}
+
+        heat_rate_GHz = _as_rate_dict(heat_rate_GHz)
+        cool_rate_GHz = _as_rate_dict(cool_rate_GHz)
+        dephase_rate_GHz = _as_rate_dict(dephase_rate_GHz)
+
+        for q, dim in qubit_dims.items():
+            gamma_h = heat_rate_GHz[q]
+            gamma_c = cool_rate_GHz[q]
+            gamma_phi = dephase_rate_GHz[q]
+
+            rate_dict[q] = _decoherence_matrix(gamma_c, gamma_phi, gamma_h, dim)
+
+        _validate_rates(qubit_dims, rate_dict)
+        self.gate_durations_ns: Dict[type, float] = gate_durations_ns
+        self.rate_matrix_GHz: Dict['cirq.Qid', np.ndarray] = rate_dict
+        self.require_physical_tag: bool = require_physical_tag
+        self.skip_measurements: bool = skip_measurements
+
+    def noisy_moment(
+        self, moment: 'cirq.Moment', system_qubits: Sequence['cirq.Qid']
+    ) -> 'cirq.OP_TREE':
+        noise_ops: List['cirq.Operation'] = []
+        moment_ns: float = 0
+        for op in moment:
+            op_duration: Optional[float] = None
+            for key, duration in self.gate_durations_ns.items():
+                if not issubclass(type(op.gate), key):
+                    continue  # gate type doesn't match
+                # TODO: remove assumption of same time across qubits
+                # if len(key) > 1 and op_data[:1] != key[:1]:
+                #     continue  # qubits don't match
+                op_duration = duration
+                break
+            if op_duration is None:
+                if not isinstance(op.gate, ops.WaitGate):
+                    continue
+                # special case for wait gates if not predefined
+                op_duration = op.gate.duration.total_nanos()
+            moment_ns = max(moment_ns, op_duration)
+
+        if moment_ns == 0:
+            return [moment]
+
+        for qubit in system_qubits:
+            qubit_op = moment.operation_at(qubit)
+            if qubit_op is None:
+                continue
+            if self.skip_measurements and protocols.is_measurement(qubit_op):
+                continue
+            if self.require_physical_tag and PHYSICAL_GATE_TAG not in qubit_op.tags:
+                # Only non-virtual gates get noise applied.
+                continue
+            rates = self.rate_matrix_GHz[qubit] * moment_ns
+            num_op = np.diag(np.sqrt(np.diag(rates)))
+            annihilation = np.sqrt(np.triu(rates, 1))
+            creation = np.sqrt(np.triu(rates.T, 1)).T
+            # Lindbladian with three Lindblad ops for the three processes
+            # Note: 'time' parameter already specified implicitly through rates
+            L = _lindbladian(annihilation) + _lindbladian(creation) + 2 * _lindbladian(num_op)
+            superop = expm(L.real)
+            kraus_ops = qis.superoperator_to_kraus(superop)
```

Comment:
Could we pull out the rest of this linear algebra into another helper (like is done above) ?

Reply:
That seems reasonable to me. Additionally I would recommend checking whether noisy_moment is unacceptably slow, in which case you could do some lru_cache tricks to significantly speed it up.

Reply:
Extracted the linear algebra to `_kraus_ops_from_rates`. Will follow up with runtime checks, but given that this is all single-qubit behavior I doubt it will be a significant issue.

Reply:
Tests reveal that this can get pretty expensive: a circuit with 10k identical ops takes ~15s to run `noisy_moments` on. I've added `lru_cache` plus necessary changes to make that work; the result takes ~0.6s on the same circuit.